### ISSUE TITLE
Forking extensions is so painful that I often avoid contributing to extensions

### ISSUE THREAD

**Issue by Yzrsah** [Microsoft/vscode#65897]
What I want to do:
- Fork an extension on GitHub
- Fix a bug

What goes wrong:
- The fork conflicts with the original extension.
- Unless I change the original id values for the extension, then I cannot use the extension manager to switch between my version and original version, because the name and metadata conflict.
- I open the terminal and manually move stuff around into enabled/disabled folders.
- I cannot see what is going on in the extension list and both extensions have the same name, I am not sure which one is enabled, is it my fork...
- So I usually end up wasting a bunch of time with updating the project and changing the publisher and changing the names to SomeExtension-MyForkedVersion, etc
- Just to fix a bug
- Sometimes I just type random values into the metadata section.
- This is a repetitive and wasteful process. So I often avoid contributing to extensions because of it.

Basically, the VSCode Extensions can't deal with forks and does not understand what they are.

The "fix":
- Modify the forked extension's package.json and change *only* the version #.
```"version": "0.0.4-myfork",```
- Then click on the extension and "Install another version..."

What goes wrong with the fix:
- "0.0.4-myfork" is a valid SemVer but it does not show up in the Extensions list. You only see "0.0.4" in the extension list, so you now have two extensions supposedly called "0.0.4" that you can switch between.

The workaround for the fix:
- The version number is not working with SemVer tag, so you must pick a new random version.
- Change to a new arbitrary version number, and hope the developer does not release a version which matches the version number you're working on.

**Comment by egamma** [comment]
@Yzrsah not clear to me in what setup you develop the forked extension.

When you fork an extension, open it in VS Code, and then run it using F5, then this version will overwrite the version of the extension that is loaded in VS Code. This is how extensions are developed and there is no need for patching needed, see https://code.visualstudio.com/api/get-started/your-first-extension.

What am I missing?

**Comment by Yzrsah** [comment]
Erich,

Just opening an extension's project and pressing F5 does not entirely help: it's an isolated action and it does not compose, therefore we're always only working on "the" extension.

Say for example:
I have 5 or 6 extensions with modifications that are currently being worked on or tested for usage in the larger scope and context of a super project with multiple project workspaces.

What is needed is a way to version these as dev-forks, rather than just invisibly loading a project for a single extension and having it replace the version with the same version that's already installed, within that single instance of VSCode only.

I need to see how the extensions work in the actual projects. I may be working on a super project extension, todo system or project manager, or a team integration. Or the functionality of the extension may be dependent on providing useful stuff that integrates with reflection or detection of code within our actual projects.

The specific thing that is needing to be tested is the functionality of the extensions as a group of betas tested *on a real project*, and on our real code.

We want it to be super simple:
- I can fork a number of dev-forks, set them to "1.0.0-foo", using the 'version' feature, and both **see** and **switch** between versions in the extension manager.

Currently there is a lot of invisibleness going on. If you replace something, you cannot see which one is really loaded, unless you alter significant meta-data. There's also a bug with the "-foo" being not displayed.

**Comment by Yzrsah** [comment]
Erich,

Just opening an extension's project and pressing F5 does not entirely help: it's an isolated action and it does not compose, therefore we're always only working on "the" extension.

Say for example:
I have 5 or 6 extensions with modifications that are currently being worked on or tested for usage in the larger scope and context of a super project with multiple project workspaces.

What is needed is a way to version these as dev-forks, rather than just invisibly loading a project for a single extension and having it replace the version with the same version that's already installed, within that single instance of VSCode only.

I need to see how the extensions work in the actual projects. I may be working on a super project extension, todo system or project manager, or a team integration. Or the functionality of the extension may be dependent on providing useful stuff that integrates with reflection or detection of code within our actual projects.

The specific thing that is needing to be tested is the functionality of the extensions as a group of betas tested on the super project, and on our real code.

We want it to be super simple:
- I can fork a number of dev-forks, set them to "1.0.0-foo", using the 'version' feature, and both **see** and **switch** between versions in the extension manager.

Currently there is a lot of invisibleness going on. If you replace something, you cannot see which one is really loaded, unless you alter significant meta-data. There's also a bug with the "-foo" being not displayed.
